In [1]:
import pandas as pd
import numpy as np

# AGGIORNAMENTO (TEMPESTIVITA')

Dimensione di qualità che valuta quanto i dati sono aggiornati rispetto al processo che li sfrutta. Nel nostro caso, laddove avevamo delle colonne con le medesime informazioni sia da API che da scraping è stata presa la decisione di mantenere quelle dello scraping, proprio perchè più aggiornate (tempestive) per il nostro obiettivo di ricerca.

# COMPLETEZZA

Vado a considerare la completezza sui dataset di API e scraping dopo l'integrazione (ossia sui dataset della cartella 'Dataset integrati 2.0').

## COMPLETEZZA POKEMON

Più che considerare la completezza a livello di intera tabella, credo sia sensato valutare colonna per colonna, dal momento che per alcuni attributi la presenza di un NaN indica che quel dato non esiste per quella specifica entità, e non che è un dato mancante (per esempio, Bulbasaur ha un NaN nella colonna 'Evolves_from' solo perchè non evolve da nessuno, non perchè manchi quella informazione).

In [2]:
pokemon = pd.read_csv('Dataset integrati 2.0/df_pokemon.csv')
pokemon

#                  Name    Species       Variant  Generation  \
0       1             Bulbasaur  Bulbasaur           NaN           1   
1       2               Ivysaur    Ivysaur           NaN           1   
2       3              Venusaur   Venusaur           NaN           1   
3       3         Mega Venusaur   Venusaur          Mega           6   
4       3   Venusaur Gigantamax   Venusaur    Gigantamax           8   
...   ...                   ...        ...           ...         ...   
1093  896             Glastrier  Glastrier           NaN           8   
1094  897             Spectrier  Spectrier           NaN           8   
1095  898               Calyrex    Calyrex           NaN           8   
1096  898  Calyrex Shadow Rider    Calyrex  Shadow Rider           8   
1097  898     Calyrex Ice Rider    Calyrex     Ice Rider           8   

         Rarity Evolves_from Has_gender_diff    Type1   Type2  ...  Attack  \
0        Common          NaN           False    Grass  Poison  ...    49.0   
1        Common    Bulbasaur           False    Grass  Poison  ...    62.0   
2        Common      Ivysaur            True    Grass  Poison  ...    82.0   
3        Common     Venusaur            True    Grass  Poison  ...   100.0   
4           NaN     Venusaur             NaN    Grass  Poison  ...     NaN   
...         ...          ...             ...      ...     ...  ...     ...   
1093  Legendary          NaN           False      Ice     NaN  ...   145.0   
1094  Legendary          NaN           False    Ghost     NaN  ...    65.0   
1095  Legendary          NaN           False  Psychic   Grass  ...    80.0   
1096  Legendary          NaN           False  Psychic   Ghost  ...    85.0   
1097  Legendary          NaN           False  Psychic     Ice  ...   165.0   

      Defense  Sp. Atk  Sp. Def  Speed  \
0        49.0     65.0     65.0   45.0   
1        63.0     80.0     80.0   60.0   
2        83.0    100.0    100.0   80.0   
3       123.0    122.0    120.0   80.0   
4         NaN      NaN      NaN    NaN   
...       ...      ...      ...    ...   
1093    130.0     65.0    110.0   30.0   
1094     60.0    145.0     80.0  130.0   
1095     80.0     80.0     80.0   80.0   
1096     80.0    165.0    100.0  150.0   
1097    150.0     85.0    130.0   50.0   

                                              image_url  \
0     https://img.pokemondb.net/sprites/sword-shield...   
1     https://img.pokemondb.net/sprites/sword-shield...   
2     https://img.pokemondb.net/sprites/sword-shield...   
3     https://img.pokemondb.net/sprites/sword-shield...   
4     https://archives.bulbagarden.net/media/upload/...   
...                                                 ...   
1093  https://img.pokemondb.net/sprites/sword-shield...   
1094  https://img.pokemondb.net/sprites/sword-shield...   
1095  https://img.pokemondb.net/sprites/sword-shield...   
1096  https://img.pokemondb.net/sprites/sword-shield...   
1097  https://img.pokemondb.net/sprites/sword-shield...   

                                          VGC2022_rules Monthly Usage (k)  \
0                                             Permitted               NaN   
1                                             Permitted               NaN   
2                                             Permitted               204   
3                                                Banned               NaN   
4                                    Gigantamax Allowed               NaN   
...                                                 ...               ...   
1093                                          Permitted                13   
1094                                          Permitted                19   
1095  Restricted (players can include two Restricted...               NaN   
1096                                                NaN               363   
1097                                                NaN               294   

     Usage Percent (%) Monthly Rank  
0                  NaN       

In [3]:
pkmn_column_nan = pd.DataFrame(pokemon.isna().sum(), columns=['Total NaN'])
pkmn_column_nan

Total NaN
#                          0
Name                       0
Species                    0
Variant                  861
Generation                 0
Rarity                    32
Evolves_from             543
Has_gender_diff           32
Type1                      0
Type2                    510
Total                     32
HP                        32
Attack                    32
Defense                   32
Sp. Atk                   32
Sp. Def                   32
Speed                     32
image_url                  0
VGC2022_rules            381
Monthly Usage (k)        756
Usage Percent (%)        756
Monthly Rank             756

In [4]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 1098 righe
pkmn_column_nan['% NaN'] = (pkmn_column_nan['Total NaN']/1098)*100
pkmn_column_nan = pkmn_column_nan.round(decimals=2)
pkmn_column_nan

Total NaN  % NaN
#                          0   0.00
Name                       0   0.00
Species                    0   0.00
Variant                  861  78.42
Generation                 0   0.00
Rarity                    32   2.91
Evolves_from             543  49.45
Has_gender_diff           32   2.91
Type1                      0   0.00
Type2                    510  46.45
Total                     32   2.91
HP                        32   2.91
Attack                    32   2.91
Defense                   32   2.91
Sp. Atk                   32   2.91
Sp. Def                   32   2.91
Speed                     32   2.91
image_url                  0   0.00
VGC2022_rules            381  34.70
Monthly Usage (k)        756  68.85
Usage Percent (%)        756  68.85
Monthly Rank             756  68.85

Sulle colonne relative alle statistiche, oltre a 'Rarity' e 'Has_gender_diff' abbiamo un 2.91% di NaN, che fanno però riferimento alle 32 varianti Gigantamax presenti nel dataset. Ci mancano quindi le informazioni di queste specifiche varianti.

Per quanto riguarda le colonne 'Variant', 'Evolves_from' e 'Type2' l'elevata percentuale di NaN non indica una mancanza di informazione, quanto più il fatto che quasi l'80% dei Pokemon non hanno varianti, circa il 50% sono Pokemon base che non evolvono da nessun altro Pokemon e poco meno del 47% dei Pokemon non hanno un secondo tipo (infatti la colonna 'Type1' non presenta NaN).

Non abbiamo informazioni per l'attributo 'VGC2022_rules' per quasi il 35% dei Pokemon. C'è poi il 68.85% di NaN nelle colonne 'Monthly Usage (k)', 'Usage Percent (%)' e 'Monthly Rank': questo dato, però, non è particolarmente preoccupante, poichè i Pokemon che per la colonna 'VGC2022_rules' hanno valore 'Banned' o un NaN non possono avere informazioni per queste tre colonne. Allo stesso modo, la presenza di NaN per gli altri Pokemon denota semplicemente come tali Pokemon non sia utilizzati in competitivo.

In [50]:
## Completezza intero dataset
print(pokemon.size)
print(pokemon.isna().sum().sum())
print(pokemon.isna().sum().sum() / pokemon.size * 100)

24156
4851
20.081967213114755


## COMPLETEZZA ITEMS

In [5]:
items = pd.read_csv('Dataset integrati 2.0/df_items.csv')
items

Name                                      Effect (alt.)  \
0    Ability Capsule  Switches a Pokémon between its two possible (n...   
1      Ability Patch                                                NaN   
2       Ability Urge   Forcibly activates a friendly Pokémon's ability.   
3          Abomasite  Held: Allows Abomasnow to Mega Evolve into Meg...   
4           Absolite  Held: Allows Absol to Mega Evolve into Mega Ab...   
..               ...                                                ...   
835     Yellow Scarf  Raises the holder's Toughness while in a contest.   
836     Yellow Shard       No effect. Can be traded for items or moves.   
837        Zap Plate  Held: Electric-Type moves from holder do 20% m...   
838             Zinc              Raises Special Defense and happiness.   
839        Zoom Lens  Held: Provides a 1/5 (20%) boost in accuracy i...   

                                                Effect  \
0    A capsule that allows a Pokémon with two Abili...   
1    Changes a Pokémon's ability to its Hidden Abil...   
2    When used, it activates the Ability of an ally...   
3      Enables Abomasnow to Mega Evolve during battle.   
4          Enables Absol to Mega Evolve during battle.   
..                                                 ...   
835         Raises holder's Tough aspect in a Contest.   
836  A small yellow shard. It appears to be from so...   
837  Increases power of Electric-type moves. Change...   
838               Increases Special Defense EVs by 10.   
839  Raises a move's accuracy if the holder moves a...   

                                      Usage Attributes Specific Category  \
0                                                  NaN          Vitamins   
1                                                  NaN          Vitamins   
2                                                  NaN   Miracle Shooter   
3                                                  NaN       Mega Stones   
4                                                  NaN       Mega Stones   
..                                                 ...               ...   
835                                ['holdable-active']           Scarves   
836                                    ['underground']      Collectibles   
837     ['holdable', 'holdable-active', 'underground']            Plates   
838  ['countable', 'consumable', 'usable-overworld'...          Vitamins   
839                    ['holdable', 'holdable-active']        Held Items   

    General Category                                          image_url  
0         Hold items  https://img.pokemondb.net/sprites/items/abilit...  
1      General items                    https://img.pokemondb.net/s.png  
2       Battle items  https://img.pokemondb.net/sprites/items/abilit...  
3         Hold items  https://img.pokemondb.net/sprites/items/abomas...  
4         Hold items  https://img.pokemondb.net/sprites/items/absoli...  
..               ...                                                ...  
835    General items  https://img.pokemondb.net/sprites/items/yellow...  
836    General items  https://img.pokemondb.net/sprites/items/yellow...  
837       Hold items  https://img.pokemondb.net/sprites/items/zap-pl...  
838         Medicine   https://img.pokemondb.net/sprites/items/zinc.png  
839       Hold items  https://img.pokemondb.net/sprites/items/zoom-l...  

[840 rows x 7 columns]

In [6]:
items_column_nan = pd.DataFrame(items.isna().sum(), columns=['Total NaN'])
items_column_nan

Total NaN
Name                       0
Effect (alt.)            224
Effect                    16
Usage Attributes         599
Specific Category          0
General Category           0
image_url                  0

In [7]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 840 righe
items_column_nan['% NaN'] = (items_column_nan['Total NaN']/840)*100
items_column_nan = items_column_nan.round(decimals=2)
items_column_nan

Total NaN  % NaN
Name                       0   0.00
Effect (alt.)            224  26.67
Effect                    16   1.90
Usage Attributes         599  71.31
Specific Category          0   0.00
General Category           0   0.00
image_url                  0   0.00

Per 16 entità del dataset Items manca l'effetto. Per poco più del 26% degli Items manca l'effetto alternativo.  La colonna 'Usage Attributes' presenta invece il 71.31% di valori mancanti: una possibile attività di improvement su tale dataset sarebbe quella di individuare informazioni più esaustive per questo attributo.

In [52]:
## Completezza intero dataset
print(items.size)
print(items.isna().sum().sum())
print(items.isna().sum().sum() / items.size * 100)

5880
839
14.268707482993198


## COMPLETEZZA MOVES

In [8]:
moves = pd.read_csv('Dataset integrati 2.0/df_moves.csv')
moves

Name  Introducted_in    Type  Power Acc.    PP Damage_class  \
0              Absorb               1   Grass   20.0  100  25.0      Special   
1          Accelerock               7    Rock   40.0  100  20.0     Physical   
2                Acid               1  Poison   40.0  100  30.0      Special   
3          Acid Armor               1  Poison    NaN  NaN  20.0       Status   
4          Acid Spray               5  Poison   40.0  100  20.0      Special   
..                ...             ...     ...    ...  ...   ...          ...   
816      G-Max Finale               8   Fairy    NaN    ∞   5.0          NaN   
817  G-Max Steelsurge               8   Steel    NaN    ∞   5.0          NaN   
818   G-Max Depletion               8  Dragon    NaN    ∞   5.0          NaN   
819    G-Max One Blow               8    Dark    NaN    ∞   5.0          NaN   
820  G-Max Rapid Flow               8   Water    NaN    ∞   5.0          NaN   

                                                Effect  Prob. (%)  
0     User recovers half the HP inflicted on opponent.        NaN  
1                                  User attacks first.        NaN  
2                May lower opponent's Special Defense.       10.0  
3                       Sharply raises user's Defense.        NaN  
4           Sharply lowers opponent's Special Defense.      100.0  
..                                                 ...        ...  
816  Heals the user and its allies by 1/6 their max...        NaN  
817  Scatters sharp spikes around the field Works l...        NaN  
818  Takes away 2 PP from the last move the target ...        NaN  
819  Hits the target even if it is protected by a p...        NaN  
820  Hits the target even if it is protected by a p...        NaN  

[821 rows x 9 columns]

In [9]:
moves_column_nan = pd.DataFrame(moves.isna().sum(), columns=['Total NaN'])
moves_column_nan

Total NaN
Name                    0
Introducted_in          0
Type                    0
Power                 346
Acc.                  216
PP                     20
Damage_class           33
Effect                 31
Prob. (%)             591

In [10]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 821 righe
moves_column_nan['% NaN'] = (moves_column_nan['Total NaN']/821)*100
moves_column_nan = moves_column_nan.round(decimals=2)
moves_column_nan

Total NaN  % NaN
Name                    0   0.00
Introducted_in          0   0.00
Type                    0   0.00
Power                 346  42.14
Acc.                  216  26.31
PP                     20   2.44
Damage_class           33   4.02
Effect                 31   3.78
Prob. (%)             591  71.99

Per circa il 4% delle mosse non abbiamo informazioni sull'attributo 'Effect' e per un altro 4% non abbiamo informazione riguardo l'attributo 'Damage_class': nel primo caso è plausibile che siano mosse senza un effetto secondario (la mossa 'Aqua Tail', per esempio, è una mossa che infligge solamente danno, senza applicare alcun effetto aggiuntivo (https://bulbapedia.bulbagarden.net/wiki/Aqua_Tail_(move))), nel secondo caso invece si tratta delle mosse G-Max che non hanno una categoria specifica di danno poichè la mossa può essere sia fisica che speciale a seconda della mossa da cui parte (https://bulbapedia.bulbagarden.net/wiki/Max_Move).

Il 72% delle mosse ha valore mancante per l'attributo 'Prob. (%)': questo però dipende dal fatto che solo alcune mosse hanno degli effetti (come può essere uno status o un drop di una statistica) che si attivano con una certa probabilità, mentre gli altri effetti sono peculiari della mossa e si attivano sempre semplicemente eseguendola.

Per quanto riguarda 'Power' la presenza dei NaN è dettata dalla presenza delle mosse status che non infliggono danni, mentre per la colonna 'Acc.' la presenza dei NaN è conseguenza del fatto che alcune mosse non possono fallire (le mosse per cui 'Acc.'=100 sono mosse che, di base, non falliscono mai, ma in presenza di un drop della precisione di un Pokemon in battaglia possono effettivamente fallire, mentre quelle con valore NaN non vengono influenzate da eventuali drop e non falliscono mai).

Abbiamo poi informazioni mancanti riguardo i PP di 20 mosse.

In [53]:
## Completezza intero dataset
print(moves.size)
print(moves.isna().sum().sum())
print(moves.isna().sum().sum() / moves.size * 100)

7389
1237
16.74110163756936


## COMPLETEZZA ABILITIES

In [11]:
ability = pd.read_csv('Dataset integrati 2.0/df_abilities.csv')
ability

Name  Generation  \
0    Adaptability           4   
1        Aerilate           6   
2       Aftermath           4   
3        Air Lock           3   
4        Analytic           5   
..            ...         ...   
262   White Smoke           3   
263      Wimp Out           7   
264  Wonder Guard           3   
265   Wonder Skin           5   
266      Zen Mode           5   

                                          Description  
0                   Powers up moves of the same type.  
1     Turns Normal-type moves into Flying-type moves.  
2     Damages the attacker landing the finishing hit.  
3                  Eliminates the effects of weather.  
4      Boosts move power when the Pokémon moves last.  
..                                                ...  
262   Prevents other Pokémon from lowering its stats.  
263            Switches out when HP drops below half.  
264               Only supereffective moves will hit.  
265  Makes status-changing moves more likely to miss.  
266            Changes form when HP drops below half.  

[267 rows x 3 columns]

In [12]:
ability_column_nan = pd.DataFrame(ability.isna().sum(), columns=['Total NaN'])
ability_column_nan

Total NaN
Name                 0
Generation           0
Description          1

In [13]:
#Calcolo la percentuale di NaN per singola colonna, considerando che abbiamo 267 righe
ability_column_nan['% NaN'] = (ability_column_nan['Total NaN']/267)*100
ability_column_nan = ability_column_nan.round(decimals=2)
ability_column_nan

Total NaN  % NaN
Name                 0   0.00
Generation           0   0.00
Description          1   0.37

Abbiamo solamente una abilità per cui manca la descrizione.

In [54]:
## Completezza intero dataset
print(ability.size)
print(ability.isna().sum().sum())
print(ability.isna().sum().sum() / ability.size * 100)

801
1
0.12484394506866417


## COMPLETEZZA NATURES

In [14]:
nature = pd.read_csv('Dataset integrati 2.0/df_natures.csv')
nature

Nature Increases Decreases Likes_berrie Dislikes_berrie
0   Adamant    Attack   Sp. Atk        Spicy             Dry
1   Bashful   Sp. Atk   Sp. Atk          Dry             Dry
2      Bold   Defense    Attack         Sour           Spicy
3     Brave    Attack     Speed        Spicy           Sweet
4      Calm   Sp. Def    Attack       Bitter           Spicy
5   Careful   Sp. Def   Sp. Atk       Bitter             Dry
6    Docile   Defense   Defense         Sour            Sour
7    Gentle   Sp. Def   Defense       Bitter            Sour
8     Hardy    Attack    Attack        Spicy           Spicy
9     Hasty     Speed   Defense        Sweet            Sour
10   Impish   Defense   Sp. Atk         Sour             Dry
11    Jolly     Speed   Sp. Atk        Sweet             Dry
12      Lax   Defense   Sp. Def         Sour          Bitter
13   Lonely    Attack   Defense        Spicy            Sour
14     Mild   Sp. Atk   Defense          Dry            Sour
15   Modest   Sp. Atk    Attack          Dry           Spicy
16    Naive     Speed   Sp. Def        Sweet          Bitter
17  Naughty    Attack   Sp. Def        Spicy          Bitter
18    Quiet   Sp. Atk     Speed          Dry           Sweet
19   Quirky   Sp. Def   Sp. Def       Bitter          Bitter
20     Rash   Sp. Atk   Sp. Def          Dry          Bitter
21  Relaxed   Defense     Speed         Sour           Sweet
22    Sassy   Sp. Def     Speed       Bitter           Sweet
23  Serious     Speed     Speed        Sweet           Sweet
24    Timid     Speed    Attack        Sweet           Spicy

In [15]:
natures_column_nan = pd.DataFrame(nature.isna().sum(), columns=['Total NaN'])
natures_column_nan

Total NaN
Nature                   0
Increases                0
Decreases                0
Likes_berrie             0
Dislikes_berrie          0

Non ci sono NaN nel dataset delle nature.

## COMPLETEZZA TYPES

In [16]:
types = pd.read_csv('Dataset integrati 2.0/df_types.csv')
types

Name  Generation
0     normal           1
1   fighting           1
2     flying           1
3     poison           1
4     ground           1
5       rock           1
6        bug           1
7      ghost           1
8      steel           2
9       fire           1
10     water           1
11     grass           1
12  electric           1
13   psychic           1
14       ice           1
15    dragon           1
16      dark           2
17     fairy           6
18   unknown           2
19    shadow           3

In [17]:
types_column_nan = pd.DataFrame(types.isna().sum(), columns=['Total NaN'])
types_column_nan

Total NaN
Name                0
Generation          0

Non ci sono NaN nel dataset dei tipi.

# CONSISTENZA

Dimensione di qualità che valuta la consistenza dei dati rispetto ai vincoli di integrità dello schema.

Rispetto alla tabella Pokemon, mi sono chiesto:

* il numero di Pokemon introdotto in ciascuna generazione è coerente col dato reale? 
* il dato totale delle statistiche di ciascun Pokemon è effettivamente uguale alla somma di tutte le sue statistiche singole?

### Punto 1

Calcolo il numero di Pokemon introdotti in ciascuna generazione, escludendo le varianti:

In [18]:
# Raggruppo per ID del Pokedex nazionale e tengo solo la prima riga di ciascun raggruppamento dopo aver ordinato per
# generazione, cioè quella riferita alla prima 'forma base' del Pokemon
pokemon_w_no_variants = pokemon.sort_values(['#','Generation']).groupby('#', as_index=False).first()
pokemon_w_no_variants

#        Name     Species       Variant  Generation     Rarity  \
0      1   Bulbasaur   Bulbasaur          None           1     Common   
1      2     Ivysaur     Ivysaur          None           1     Common   
2      3    Venusaur    Venusaur          Mega           1     Common   
3      4  Charmander  Charmander          None           1     Common   
4      5  Charmeleon  Charmeleon          None           1     Common   
..   ...         ...         ...           ...         ...        ...   
893  894   Regieleki   Regieleki          None           8  Legendary   
894  895   Regidrago   Regidrago          None           8  Legendary   
895  896   Glastrier   Glastrier          None           8  Legendary   
896  897   Spectrier   Spectrier          None           8  Legendary   
897  898     Calyrex     Calyrex  Shadow Rider           8  Legendary   

    Evolves_from  Has_gender_diff     Type1   Type2  ...  Attack  Defense  \
0           None            False     Grass  Poison  ...    49.0     49.0   
1      Bulbasaur            False     Grass  Poison  ...    62.0     63.0   
2        Ivysaur             True     Grass  Poison  ...    82.0     83.0   
3           None            False      Fire    None  ...    52.0     43.0   
4     Charmander            False      Fire    None  ...    64.0     58.0   
..           ...              ...       ...     ...  ...     ...      ...   
893         None            False  Electric    None  ...   100.0     50.0   
894         None            False    Dragon    None  ...   100.0     50.0   
895         None            False       Ice    None  ...   145.0    130.0   
896         None            False     Ghost    None  ...    65.0     60.0   
897         None            False   Psychic   Grass  ...    80.0     80.0   

     Sp. Atk  Sp. Def  Speed  \
0       65.0     65.0   45.0   
1       80.0     80.0   60.0   
2      100.0    100.0   80.0   
3       60.0     50.0   65.0   
4       80.0     65.0   80.0   
..       ...      ...    ...   
893    100.0     50.0  200.0   
894    100.0     50.0   80.0   
895     65.0    110.0   30.0   
896    145.0     80.0  130.0   
897     80.0     80.0   80.0   

                                             image_url  \
0    https://img.pokemondb.net/sprites/sword-shield...   
1    https://img.pokemondb.net/sprites/sword-shield...   
2    https://img.pokemondb.net/sprites/sword-shield...   
3    https://img.pokemondb.net/sprites/sword-shield...   
4    https://img.pokemondb.net/sprites/sword-shield...   
..                                                 ...   
893  https://img.pokemondb.net/sprites/sword-shield...   
894  https://img.pokemondb.net/sprites/sword-shield...   
895  https://img.pokemondb.net/sprites/sword-shield...   
896  https://img.pokemondb.net/sprites/sword-shield...   
897  https://img.pokemondb.net/sprites/sword-shield...   

                                         VGC2022_rules Monthly Usage (k)  \
0                                            Permitted              None   
1                                            Permitted              None   
2                                            Permitted               204   
3                                            Permitted              None   
4                                            Permitted              None   
..                                                 ...               ...   
893                                          Permitted               603   
894                                          Permitted                12   
895                                          Permitted                13   
896                                          Permitted                19   
897  Restricted (players can include two Restricted...               363   

    Usage Percent (%) Monthly Rank  
0                None         None  
1                None         None  
2                   7           24  
3                None         None  
4                None    

In [19]:
n_pokemon_by_generation = {
  1: 151,
  2: 100,
  3: 135,
  4: 107,
  5: 156,
  6: 72,
  7: 88,
  8: 89
}

In [20]:
n_pokemon_by_generation

{1: 151, 2: 100, 3: 135, 4: 107, 5: 156, 6: 72, 7: 88, 8: 89}

In [21]:
df_pokemon_per_gen = dict(pokemon_w_no_variants.groupby('Generation').size())
df_pokemon_per_gen

{1: 151, 2: 100, 3: 135, 4: 107, 5: 156, 6: 72, 7: 88, 8: 89}

In [22]:
df_pokemon_per_gen == n_pokemon_by_generation

True

Le informazioni circa la generazione di introduzione di ciascuna specie sono consistenti con i dati reali.

### Punto 2

In [23]:
mask = pokemon['Total'] != pokemon['HP']+pokemon['Attack']+pokemon['Defense']+pokemon['Sp. Atk']+pokemon['Sp. Def']+pokemon['Speed']
columns = ['#', 'Name', 'Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']
pokemon[columns][(mask)]

#                                Name  Total  HP  Attack  Defense  \
4       3                 Venusaur Gigantamax    NaN NaN     NaN      NaN   
8       6                Charizard Gigantamax    NaN NaN     NaN      NaN   
13      9                Blastoise Gigantamax    NaN NaN     NaN      NaN   
18     12               Butterfree Gigantamax    NaN NaN     NaN      NaN   
36     25                  Pikachu Gigantamax    NaN NaN     NaN      NaN   
75     52                   Meowth Gigantamax    NaN NaN     NaN      NaN   
95     68                  Machamp Gigantamax    NaN NaN     NaN      NaN   
134    94                   Gengar Gigantamax    NaN NaN     NaN      NaN   
140    99                  Kingler Gigantamax    NaN NaN     NaN      NaN   
182   131                   Lapras Gigantamax    NaN NaN     NaN      NaN   
186   133                    Eevee Gigantamax    NaN NaN     NaN      NaN   
198   143                  Snorlax Gigantamax    NaN NaN     NaN      NaN   
699   569                 Garbodor Gigantamax    NaN NaN     NaN      NaN   
979   809                 Melmetal Gigantamax    NaN NaN     NaN      NaN   
983   812                Rillaboom Gigantamax    NaN NaN     NaN      NaN   
987   815                Cinderace Gigantamax    NaN NaN     NaN      NaN   
991   818                 Inteleon Gigantamax    NaN NaN     NaN      NaN   
998   823              Corviknight Gigantamax    NaN NaN     NaN      NaN   
1002  826                 Orbeetle Gigantamax    NaN NaN     NaN      NaN   
1010  834                  Drednaw Gigantamax    NaN NaN     NaN      NaN   
1016  839                Coalossal Gigantamax    NaN NaN     NaN      NaN   
1020  841                  Flapple Gigantamax    NaN NaN     NaN      NaN   
1022  842                 Appletun Gigantamax    NaN NaN     NaN      NaN   
1024  844               Sandaconda Gigantamax    NaN NaN     NaN      NaN   
1033  851              Centiskorch Gigantamax    NaN NaN     NaN      NaN   
1042  858                Hatterene Gigantamax    NaN NaN     NaN      NaN   
1045  861               Grimmsnarl Gigantamax    NaN NaN     NaN      NaN   
1055  869                 Alcremie Gigantamax    NaN NaN     NaN      NaN   
1068  879               Copperajah Gigantamax    NaN NaN     NaN      NaN   
1075  884                Duraludon Gigantamax    NaN NaN     NaN      NaN   
1086  892  Urshifu (Single Strike) Gigantamax    NaN NaN     NaN      NaN   
1087  892   Urshifu (Rapid Strike) Gigantamax    NaN NaN     NaN      NaN   

      Sp. Atk  Sp. Def  Speed  
4         NaN      NaN    NaN  
8         NaN      NaN    NaN  
13        NaN      NaN    NaN  
18        NaN      NaN    NaN  
36        NaN      NaN    NaN  
75        NaN      NaN    NaN  
95        NaN      NaN    NaN  
134       NaN      NaN    NaN  
140       NaN      NaN    NaN  
182       NaN      NaN    NaN  
186       NaN      NaN    NaN  
198       NaN      NaN    NaN  
699       NaN      NaN    NaN  
979       NaN      NaN    NaN  
983       NaN      NaN    NaN  
987       NaN      NaN    NaN  
991       NaN      NaN    NaN  
998       NaN      NaN    NaN  
1002      NaN      NaN    NaN  
1010      NaN      NaN    NaN  
1016      NaN      NaN    NaN  
1020      NaN      NaN    NaN  
1022      NaN      NaN    NaN  
1024      NaN      NaN    NaN  
1033      NaN      NaN    NaN  
1042      NaN      NaN    NaN  
1045      NaN      NaN    NaN  
1055      NaN      NaN    NaN  
1068      NaN      NaN    NaN  
1075      NaN      NaN    NaN  
1086      NaN      NaN    NaN  
1087      NaN      NaN    NaN

Tutte le righe del dataset Pokemon, ad eccezione di quelle per le forme Gigantamax di cui mancano le statistiche di base, hanno l'attributo 'Total' pari alla somma delle singole statistiche.

# RIDONDANZA

Dimensione di qualità che valuta il numero di righe duplicate (più precisamente, riferite alla stessa entità) presente nelle varie tabelle.

Vado a valutarla rispetto all'attributo identificativo delle varie tabelle. Viene fatto sia sulle tabelle di partenza (dopo averle sistemate) che sulle tabelle finali, post integrazione.

### POKEMON SCRAPING

In [24]:
pokedex_scraping = pd.read_json('Scraping_datasets/pokemondb/pokedex.json')
pokedex_scraping

#         Name   Type1     Type2  Total   HP  Attack  Defense  \
0       1    Bulbasaur   Grass    Poison    318   45      49       49   
1       2      Ivysaur   Grass    Poison    405   60      62       63   
2       3     Venusaur   Grass    Poison    525   80      82       83   
3       3     Venusaur   Grass    Poison    625   80     100      123   
4       4   Charmander    Fire       NaN    309   39      52       43   
...   ...          ...     ...       ...    ...  ...     ...      ...   
1070  902  Basculegion   Water     Ghost    530  120      92       65   
1071  903     Sneasler  Poison  Fighting    510   80     130       60   
1072  904     Overqwil    Dark    Poison    510   85     115       95   
1073  905     Enamorus   Fairy    Flying    580   74     115       70   
1074  905     Enamorus   Fairy    Flying    580   74     115      110   

      Sp. Atk  Sp. Def  Speed  \
0          65       65     45   
1          80       80     60   
2         100      100     80   
3         122      120     80   
4          60       50     65   
...       ...      ...    ...   
1070      100       75     78   
1071       40       80    120   
1072       65       65     85   
1073      135       80    106   
1074      135      100     46   

                                             sprite_url          Variant  
0     https://img.pokemondb.net/sprites/sword-shield...              NaN  
1     https://img.pokemondb.net/sprites/sword-shield...              NaN  
2     https://img.pokemondb.net/sprites/sword-shield...              NaN  
3     https://img.pokemondb.net/sprites/sword-shield...    Mega Venusaur  
4     https://img.pokemondb.net/sprites/sword-shield...              NaN  
...                                                 ...              ...  
1070                    https://img.pokemondb.net/s.png           Female  
1071                    https://img.pokemondb.net/s.png              NaN  
1072                    https://img.pokemondb.net/s.png              NaN  
1073                    https://img.pokemondb.net/s.png  Incarnate Forme  
1074                    https://img.pokemondb.net/s.png    Therian Forme  

[1075 rows x 13 columns]

In [25]:
redundancy_pokemon_scraping = pokedex_scraping.groupby(['#', 'Name'], as_index=False).size()
redundancy_pokemon_scraping = redundancy_pokemon_scraping[redundancy_pokemon_scraping['size']>1]
redundancy_pokemon_scraping

#         Name  size
2      3     Venusaur     2
5      6    Charizard     3
8      9    Blastoise     2
14    15     Beedrill     2
17    18      Pidgeot     2
..   ...          ...   ...
889  890    Eternatus     2
891  892      Urshifu     2
897  898      Calyrex     3
901  902  Basculegion     2
904  905     Enamorus     2

[141 rows x 3 columns]

La ridondanza è legata alla presenza di diverse righe per le diverse varianti dello stesso Pokemon.

### POKEDEX API

In [26]:
pokedex_api = pd.read_json('API_per_integrazione/species.json.gz', compression='gzip', orient='records', lines=True)
pokedex_api

ID        Name                                          Varieties  \
0      1   Bulbasaur  [{'is_default': True, 'pokemon': {'name': 'bul...   
1      2     Ivysaur  [{'is_default': True, 'pokemon': {'name': 'ivy...   
2      3    Venusaur  [{'is_default': True, 'pokemon': {'name': 'ven...   
3      4  Charmander  [{'is_default': True, 'pokemon': {'name': 'cha...   
4      5  Charmeleon  [{'is_default': True, 'pokemon': {'name': 'cha...   
..   ...         ...                                                ...   
893  894   Regieleki  [{'is_default': True, 'pokemon': {'name': 'reg...   
894  895   Regidrago  [{'is_default': True, 'pokemon': {'name': 'reg...   
895  896   Glastrier  [{'is_default': True, 'pokemon': {'name': 'gla...   
896  897   Spectrier  [{'is_default': True, 'pokemon': {'name': 'spe...   
897  898     Calyrex  [{'is_default': True, 'pokemon': {'name': 'cal...   

     Generation Evolves_from  Has_gender_diff     Rarity  
0             1         None            False     Common  
1             1    Bulbasaur            False     Common  
2             1      Ivysaur             True     Common  
3             1         None            False     Common  
4             1   Charmander            False     Common  
..          ...          ...              ...        ...  
893           8         None            False  Legendary  
894           8         None            False  Legendary  
895           8         None            False  Legendary  
896           8         None            False  Legendary  
897           8         None            False  Legendary  

[898 rows x 7 columns]

In [27]:
redundancy_pokemon_api = pokedex_api.groupby(['ID', 'Name'], as_index=False).size()
redundancy_pokemon_api = redundancy_pokemon_api[redundancy_pokemon_api['size']>1]
redundancy_pokemon_api

Empty DataFrame
Columns: [ID, Name, size]
Index: []

Non ci sono righe duplicate per quanto riguarda l'API.

### ITEMS SCRAPING

In [28]:
items_scraping = pd.read_json('Scraping_datasets/pokemondb/items.json')
items_scraping

Name       Category  \
0    Ability Capsule     Hold items   
1      Ability Patch  General items   
2       Ability Urge   Battle items   
3          Abomasite     Hold items   
4           Absolite     Hold items   
..               ...            ...   
955     Yellow Scarf  General items   
956     Yellow Shard  General items   
957        Zap Plate     Hold items   
958             Zinc       Medicine   
959        Zoom Lens     Hold items   

                                                Effect  \
0    A capsule that allows a Pokémon with two Abili...   
1    Changes a Pokémon's ability to its Hidden Abil...   
2    When used, it activates the Ability of an ally...   
3      Enables Abomasnow to Mega Evolve during battle.   
4          Enables Absol to Mega Evolve during battle.   
..                                                 ...   
955         Raises holder's Tough aspect in a Contest.   
956  A small yellow shard. It appears to be from so...   
957  Increases power of Electric-type moves. Change...   
958               Increases Special Defense EVs by 10.   
959  Raises a move's accuracy if the holder moves a...   

                                            sprite_url  
0    https://img.pokemondb.net/sprites/items/abilit...  
1                      https://img.pokemondb.net/s.png  
2    https://img.pokemondb.net/sprites/items/abilit...  
3    https://img.pokemondb.net/sprites/items/abomas...  
4    https://img.pokemondb.net/sprites/items/absoli...  
..                                                 ...  
955  https://img.pokemondb.net/sprites/items/yellow...  
956  https://img.pokemondb.net/sprites/items/yellow...  
957  https://img.pokemondb.net/sprites/items/zap-pl...  
958   https://img.pokemondb.net/sprites/items/zinc.png  
959  https://img.pokemondb.net/sprites/items/zoom-l...  

[960 rows x 4 columns]

In [29]:
redundancy_items_scraping = items_scraping.groupby(['Name'], as_index=False).size()
redundancy_items_scraping = redundancy_items_scraping[redundancy_items_scraping['size']>1]
redundancy_items_scraping

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### ITEMS API

In [30]:
items_api = pd.read_json('API_per_integrazione/items.json.gz', compression='gzip', orient='records', lines=True)
items_api

ID              Name  \
0        1       Master-ball   
1        2        Ultra-ball   
2        3        Great-ball   
3        4         Poke-ball   
4        5       Safari-ball   
...    ...               ...   
1602  1603  Shaderoot-carrot   
1603  1604        Dynite-ore   
1604  1605      Carrot-seeds   
1605  1606     Ability-patch   
1606  1607    Reins-of-unity   

                                             Attributes          Category  \
0     [countable, consumable, usable-in-battle, hold...    Standard-balls   
1     [countable, consumable, usable-in-battle, hold...    Standard-balls   
2     [countable, consumable, usable-in-battle, hold...    Standard-balls   
3     [countable, consumable, usable-in-battle, hold...    Standard-balls   
4     [countable, consumable, usable-in-battle, hold...    Standard-balls   
...                                                 ...               ...   
1602                                                 []  Plot-advancement   
1603                                                 []      Collectibles   
1604                                                 []  Plot-advancement   
1605                                                 []          Vitamins   
1606                                                 []  Plot-advancement   

                                                 Effect  
0                    Catches a wild Pokémon every time.  
1     Tries to catch a wild Pokémon.  Success rate i...  
2     Tries to catch a wild Pokémon.  Success rate i...  
3                        Tries to catch a wild Pokémon.  
4     Tries to catch a wild Pokémon in the Great Mar...  
...                                                 ...  
1602                                               None  
1603                                               None  
1604                                               None  
1605                                               None  
1606                                               None  

[1607 rows x 5 columns]

In [31]:
redundancy_items_api = items_api.groupby(['Name'], as_index=False).size()
redundancy_items_api = redundancy_items_api[redundancy_items_api['size']>1]
redundancy_items_api

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### ABILITIES SCRAPING

In [32]:
ability_scraping = pd.read_json('Scraping_datasets/pokemondb/abilities.json')
ability_scraping

Name  Pokémon                                       Description  \
0    Adaptability       16                 Powers up moves of the same type.   
1        Aerilate        2   Turns Normal-type moves into Flying-type moves.   
2       Aftermath       10   Damages the attacker landing the finishing hit.   
3        Air Lock        1                Eliminates the effects of weather.   
4        Analytic       12    Boosts move power when the Pokémon moves last.   
..            ...      ...                                               ...   
261   White Smoke        4   Prevents other Pokémon from lowering its stats.   
262      Wimp Out        1            Switches out when HP drops below half.   
263  Wonder Guard        1               Only supereffective moves will hit.   
264   Wonder Skin        5  Makes status-changing moves more likely to miss.   
265      Zen Mode        4            Changes form when HP drops below half.   

     Gen.  
0       4  
1       6  
2       4  
3       3  
4       5  
..    ...  
261     3  
262     7  
263     3  
264     5  
265     5  

[266 rows x 4 columns]

In [33]:
redundancy_ability_scraping = ability_scraping.groupby(['Name'], as_index=False).size()
redundancy_ability_scraping = redundancy_ability_scraping[redundancy_ability_scraping['size']>1]
redundancy_ability_scraping

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### ABILITIES API

In [34]:
ability_api = pd.read_json('API_per_integrazione/Abilities.json.gz', compression='gzip', orient='records', lines=True)
ability_api

ID              Name  Generation  Is_Main_Series  \
0      1            Stench           3            True   
1      2           Drizzle           3            True   
2      3       Speed-boost           3            True   
3      4      Battle-armor           3            True   
4      5            Sturdy           3            True   
..   ...               ...         ...             ...   
262  263       Dragons-maw           8            True   
263  264    Chilling-neigh           8            True   
264  265        Grim-neigh           8            True   
265  266  As-one-glastrier           8            True   
266  267  As-one-spectrier           8            True   

                                        Effect_changes  \
0    [{'effect_entries': [{'effect': 'Hat im Kampf ...   
1                                                   []   
2                                                   []   
3                                                   []   
4    [{'effect_entries': [{'effect': 'Verhindert ke...   
..                                                 ...   
262                                                 []   
263                                                 []   
264                                                 []   
265                                                 []   
266                                                 []   

                                        Effect_entries  \
0    This Pokémon's damaging moves have a 10% chanc...   
1    The weather changes to rain when this Pokémon ...   
2    This Pokémon's Speed rises one stage after eac...   
3    Moves cannot score critical hits against this ...   
4    When this Pokémon is at full HP, any hit that ...   
..                                                 ...   
262                                               None   
263                                               None   
264                                               None   
265                                               None   
266                                               None   

                                               Pokemon  
0    [{'is_hidden': True, 'pokemon': {'name': 'gloo...  
1    [{'is_hidden': True, 'pokemon': {'name': 'poli...  
2    [{'is_hidden': False, 'pokemon': {'name': 'yan...  
3    [{'is_hidden': True, 'pokemon': {'name': 'cubo...  
4    [{'is_hidden': False, 'pokemon': {'name': 'geo...  
..                                                 ...  
262  [{'is_hidden': False, 'pokemon': {'name': 'reg...  
263  [{'is_hidden': False, 'pokemon': {'name': 'gla...  
264  [{'is_hidden': False, 'pokemon': {'name': 'spe...  
265  [{'is_hidden': False, 'pokemon': {'name': 'cal...  
266  [{'is_hidden': False, 'pokemon': {'name': 'cal...  

[267 rows x 7 columns]

In [35]:
redundancy_ability_api = ability_api.groupby(['Name'], as_index=False).size()
redundancy_ability_api = redundancy_ability_api[redundancy_ability_api['size']>1]
redundancy_ability_api

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### MOVES SCRAPING

In [36]:
moves_scraping = pd.read_json('Scraping_datasets/pokemondb/moves.json')
moves_scraping

Name      Type Power Acc.  PP  \
0    10,000,000 Volt Thunderbolt  Electric   195    —   1   
1                         Absorb     Grass    20  100  25   
2                     Accelerock      Rock    40  100  20   
3                           Acid    Poison    40  100  30   
4                     Acid Armor    Poison     —    —  20   
..                           ...       ...   ...  ...  ..   
860                         Yawn    Normal     —    —  10   
861                   Zap Cannon  Electric   120   50   5   
862                 Zen Headbutt   Psychic    80   90  15   
863                     Zing Zap  Electric    80  100  10   
864                    Zippy Zap  Electric    50  100  15   

                                                Effect Prob. (%) Cat.  
0    Pikachu-exclusive Z-Move. High critical hit ra...         —  NaN  
1     User recovers half the HP inflicted on opponent.         —  NaN  
2                                  User attacks first.         —  NaN  
3                May lower opponent's Special Defense.        10  NaN  
4                       Sharply raises user's Defense.         —  NaN  
..                                                 ...       ...  ...  
860           Puts opponent to sleep in the next turn.         —  NaN  
861                                Paralyzes opponent.       100  NaN  
862                               May cause flinching.        20  NaN  
863                               May cause flinching.        30  NaN  
864                  Always results in a critical hit.       100  NaN  

[865 rows x 8 columns]

In [37]:
redundancy_moves_scraping = moves_scraping.groupby(['Name'], as_index=False).size()
redundancy_moves_scraping = redundancy_moves_scraping[redundancy_moves_scraping['size']>1]
redundancy_moves_scraping

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### MOVES API

In [38]:
moves_api = pd.read_json('API_per_integrazione/moves.json.gz', compression='gzip', orient='records', lines=True)
moves_api

ID          Name      Type  Power  Accuracy    PP Damage_class  \
0      1         Pound    Normal   40.0     100.0  35.0     Physical   
1      2   Karate-chop  Fighting   50.0     100.0  25.0     Physical   
2      3   Double-slap    Normal   15.0      85.0  10.0     Physical   
3      4   Comet-punch    Normal   18.0      85.0  15.0     Physical   
4      5    Mega-punch    Normal   80.0      85.0  20.0     Physical   
..   ...           ...       ...    ...       ...   ...          ...   
839  840   Shadow-hold    Shadow    NaN       NaN   NaN       Status   
840  841   Shadow-mist    Shadow    NaN     100.0   NaN       Status   
841  842  Shadow-panic    Shadow    NaN      90.0   NaN       Status   
842  843   Shadow-shed    Shadow    NaN       NaN   NaN       Status   
843  844    Shadow-sky    Shadow    NaN       NaN   NaN       Status   

     Introducted_in                                         Learned_by  
0                 1  [{'name': 'clefairy', 'url': 'https://pokeapi....  
1                 1  [{'name': 'mankey', 'url': 'https://pokeapi.co...  
2                 1  [{'name': 'clefairy', 'url': 'https://pokeapi....  
3                 1  [{'name': 'hitmonchan', 'url': 'https://pokeap...  
4                 1  [{'name': 'charmander', 'url': 'https://pokeap...  
..              ...                                                ...  
839               3                                                 []  
840               3                                                 []  
841               3                                                 []  
842               3                                                 []  
843               3                                                 []  

[844 rows x 9 columns]

In [39]:
redundancy_moves_api = moves_api.groupby(['Name'], as_index=False).size()
redundancy_moves_api = redundancy_moves_api[redundancy_moves_api['size']>1]
redundancy_moves_api

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### TYPES SCRAPING

In [40]:
redundancy_types = types.groupby(['Name'], as_index=False).size()
redundancy_types = redundancy_types[redundancy_types['size']>1]
redundancy_types

Empty DataFrame
Columns: [Name, size]
Index: []

Non ci sono elementi duplicati.

### NATURES SCRAPING

In [41]:
redundancy_natures = nature.groupby(['Nature'], as_index=False).size()
redundancy_natures = redundancy_natures[redundancy_natures['size']>1]
redundancy_natures

Empty DataFrame
Columns: [Nature, size]
Index: []

Non ci sono elementi duplicati.

Non ho valutato gli altri dataset con cui sono stati integrati i dati del competitivo, perchè credo che come struttura questa analisi sia da fare giusto sulle entità del database.

# ACCURATEZZA

Dimensione di qualità che valuta quanto un valore è vicino alla rappresentazione reale dell'oggetto che rappresenta (es. quanto il nome del Pokemon è vicino al nome reale del Pokemon stesso). Può essere valutata, a livello sintattico, in termini di Edit Distance.

### Pokemon

In [42]:
from nltk.metrics import edit_distance

In [43]:
pokemon_api_names = pokedex_api[['ID', 'Name']]
pokemon_api_names.head()

ID        Name
0   1   Bulbasaur
1   2     Ivysaur
2   3    Venusaur
3   4  Charmander
4   5  Charmeleon

In [44]:
pokemon_scrap_names = pokedex_scraping.groupby('#', as_index=False)['Name'].first()
pokemon_scrap_names.columns = ['#', 'Reference']
pokemon_scrap_names.head()

#   Reference
0  1   Bulbasaur
1  2     Ivysaur
2  3    Venusaur
3  4  Charmander
4  5  Charmeleon

In [45]:
pokemon_names = pd.merge(pokemon_api_names, pokemon_scrap_names, left_on='ID', right_on='#')
pokemon_names.drop('#', axis=1, inplace=True)
pokemon_names

ID        Name   Reference
0      1   Bulbasaur   Bulbasaur
1      2     Ivysaur     Ivysaur
2      3    Venusaur    Venusaur
3      4  Charmander  Charmander
4      5  Charmeleon  Charmeleon
..   ...         ...         ...
893  894   Regieleki   Regieleki
894  895   Regidrago   Regidrago
895  896   Glastrier   Glastrier
896  897   Spectrier   Spectrier
897  898     Calyrex     Calyrex

[898 rows x 3 columns]

In [46]:
pokemon_names["Edit Distance"] = pokemon_names.loc[:, ["Name", "Reference"]].apply(lambda x: edit_distance(*x), axis=1)
pokemon_names[pokemon_names["Edit Distance"] != 0]

ID       Name   Reference  Edit Distance
28    29  Nidoran-f    Nidoran♀              2
31    32  Nidoran-m    Nidoran♂              2
82    83  Farfetchd  Farfetch'd              1
121  122    Mr-mime         Mr.              5
438  439    Mime-jr        Mime              3
473  474  Porygon-z   Porygon-Z              1
668  669    Flabebe     Flabébé              2
771  772  Type-null       Type:              5
784  785  Tapu-koko        Tapu              5
785  786  Tapu-lele        Tapu              5
786  787  Tapu-bulu        Tapu              5
787  788  Tapu-fini        Tapu              5
864  865  Sirfetchd  Sirfetch'd              1
865  866    Mr-rime         Mr.              5

Sui Pokemon è stato possibile fare analisi dell'Edit Distance poichè abbiamo l'ID del Pokedex che individua univocamente ogni singolo Pokemon. Su items, mosse, ecc... non so bene come fare.